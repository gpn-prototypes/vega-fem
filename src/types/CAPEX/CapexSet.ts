import CapexExpenseSetGroup from './CapexExpenseSetGroup';
import CapexSetGlobalValue from './CapexSetGlobalValue';

// interface for scenarios of capex
export default interface CapexSet {
  years?: number;
  yearStart?: number;
  capexGlobalValueList: CapexSetGlobalValue[];
  capexExpenseGroupList?: CapexExpenseSetGroup[];
}
