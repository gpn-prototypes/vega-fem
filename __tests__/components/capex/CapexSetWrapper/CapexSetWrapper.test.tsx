import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { fireEvent, render, RenderResult, screen } from '@testing-library/react';

import {
  CapexSetWrapper,
  CapexSetWrapperProps,
} from '../../../../src/components/CAPEX/CapexSetWrapper/CapexSetWrapper';
import CapexSet from '../../../../types/CAPEX/CapexSet';

let fakeCapexSet: CapexSet;
// mocking table
jest.mock('../../../../src/containers/CAPEX/CapexTableContainer', () => {
  return {
    CapexTableContainer: () => {
      return <div />;
    },
  };
});

beforeEach(() => {
  fakeCapexSet = {
    years: 15,
    yearStart: 2020,
    capexGlobalValueList: [
      {
        id: '3',
        name: 'reserveValue',
        caption: 'Величина резерва',
        unit: 'rub.',
        value: 38.0,
      },
    ],
    capexExpenseGroupList: [
      {
        id: '1',
        name: 'oneTimePaymentGroup',
        caption: 'Первоначальный взнос',
        valueTotal: 1000000.0,
        capexExpenseList: [
          {
            id: '1',
            name: 'oneTimePaymentValue',
            caption: 'Значение разового платежа',
            unit: 'rub.',
            valueTotal: 1000000.0,
          },
        ],
      },
    ],
  };
});

const renderComponent = (props: CapexSetWrapperProps): RenderResult =>
  render(<CapexSetWrapper {...props} />);

describe('CapexSetWrapper', () => {
  test('Рендерится без ошибок', () => {
    expect(renderComponent).not.toThrow();
  });

  test('Рендер соответствует снимку', () => {
    const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
      .create(
        <CapexSetWrapper
          capexSet={fakeCapexSet}
          updateCapexValue={jest.fn()}
          requestDeleteCapexGroup={jest.fn()}
          requestChangeCapexGroup={jest.fn()}
          deleteCapexValue={jest.fn()}
          addCapex={jest.fn()}
          addCapexSetGroup={jest.fn()}
          highlightArticle={jest.fn()}
          highlightArticleClear={jest.fn()}
          updateCapexGlobalValue={jest.fn()}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Рендерится с правильно подставленным fakeCapexSet', () => {
    renderComponent({
      capexSet: fakeCapexSet,
      addCapex: jest.fn(),
      addCapexSetGroup: jest.fn(),
      deleteCapexValue: jest.fn(),
      highlightArticle: jest.fn(),
      highlightArticleClear: jest.fn(),
      requestChangeCapexGroup: jest.fn(),
      requestDeleteCapexGroup: jest.fn(),
      updateCapexGlobalValue: jest.fn(),
      updateCapexValue: jest.fn(),
    });

    expect(screen.getByDisplayValue('38')).toBeInTheDocument();
  });

  test('срабатывают ивенты на изменение глобального значения', () => {
    const fakeUpdateCapexGlobalValue = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        updateCapexValue={jest.fn()}
        requestDeleteCapexGroup={jest.fn()}
        requestChangeCapexGroup={jest.fn()}
        deleteCapexValue={jest.fn()}
        addCapex={jest.fn()}
        addCapexSetGroup={jest.fn()}
        highlightArticle={jest.fn()}
        highlightArticleClear={jest.fn()}
        updateCapexGlobalValue={fakeUpdateCapexGlobalValue}
      />,
    );
    const globalValueInput: HTMLElement = screen.getByDisplayValue('38');
    fireEvent.change(globalValueInput, { target: { value: '50' } });
    expect(globalValueInput).toHaveAttribute('value', '50');
    fireEvent.blur(globalValueInput);
    expect(fakeUpdateCapexGlobalValue).toBeCalledTimes(1);
  });

  test('срабатывают ивенты на добавление группы', () => {
    const fakeAddGroup = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        updateCapexValue={jest.fn()}
        requestDeleteCapexGroup={jest.fn()}
        requestChangeCapexGroup={jest.fn()}
        deleteCapexValue={jest.fn()}
        addCapex={jest.fn()}
        addCapexSetGroup={fakeAddGroup}
        highlightArticle={jest.fn()}
        highlightArticleClear={jest.fn()}
        updateCapexGlobalValue={jest.fn()}
      />,
    );

    const addToggle: HTMLElement = screen.getByText(/Добавить группу затрат/i);
    fireEvent.click(addToggle);
    const addGroupInput: HTMLElement = screen.getByPlaceholderText(
      'Введите название группы затрат',
    );
    const addGroupButton: HTMLElement = screen.getByText('Добавить группу');

    expect(addGroupInput).toBeInTheDocument();
    fireEvent.change(addGroupInput, { target: { value: 'new group' } });
    fireEvent.click(addGroupButton);
    expect(fakeAddGroup).toBeCalledTimes(1);
  });

  test('отмена добавления группы', () => {
    const fakeAddGroup = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        updateCapexValue={jest.fn()}
        requestDeleteCapexGroup={jest.fn()}
        requestChangeCapexGroup={jest.fn()}
        deleteCapexValue={jest.fn()}
        addCapex={jest.fn()}
        addCapexSetGroup={fakeAddGroup}
        highlightArticle={jest.fn()}
        highlightArticleClear={jest.fn()}
        updateCapexGlobalValue={jest.fn()}
      />,
    );

    const addToggle: HTMLElement = screen.getByText(/добавить группу затрат/i);
    fireEvent.click(addToggle);
    const addGroupInput: HTMLElement = screen.getByPlaceholderText(
      'Введите название группы затрат',
    );
    const addGroupCancelButton: HTMLElement = screen.getByText('Отмена');

    expect(addGroupInput).toBeInTheDocument();
    fireEvent.change(addGroupInput, { target: { value: 'new group' } });
    fireEvent.click(addGroupCancelButton);
    expect(fakeAddGroup).toBeCalledTimes(0);
  });
});
