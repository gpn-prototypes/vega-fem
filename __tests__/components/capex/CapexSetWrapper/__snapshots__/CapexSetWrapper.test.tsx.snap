// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CapexSetWrapper Рендер соответствует снимку 1`] = `
<div
  className="BlockWrapper"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <span
        className="Text Text_size_xs Text_view_primary"
      >
        CAPEX
      </span>
      <span
        className="Icon Icon_size_xs IconSelect"
      >
        <svg
          className="Icon-Svg"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 8.5L10 4H2l4 4.5z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <form
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      onSubmit={[Function]}
    >
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <div
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          <div
            className={
              Object {
                "has": [Function],
                "is": [Function],
                "mix": [Function],
                "split": [Function],
                "state": [Function],
                "toString": [Function],
              }
            }
          >
            <label
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              htmlFor="capexSetreserveValue"
            >
              Величина резерва
            </label>
            <div
              className="TextField TextField_size_s TextField_view_default TextField_form_default TextField_width_full TextField_type_text TextField_withValue"
              onKeyDown={[Function]}
            >
              <input
                autoFocus={false}
                className="TextField-Input"
                id="capexSetreserveValue"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value="38"
              />
              <div
                className="TextField-Side TextField-Side_position_right TextField-Side_type_string"
                title="rub."
              >
                rub.
              </div>
            </div>
          </div>
        </div>
        <div
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          <div
            className="GroupWrapper"
          >
            <div
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
                onClick={[Function]}
                role="presentation"
              >
                <span
                  className="Icon Icon_size_xs IconArrowDown"
                >
                  <svg
                    className="Icon-Svg"
                    viewBox="0 0 12 12"
                  >
                    <path
                      clipRule="evenodd"
                      d="M6.005 8.341L2.03 4.366l.707-.707 3.268 3.268 3.268-3.268.707.707-3.975 3.975z"
                      fillRule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  className="Text Text_size_s Text_view_primary"
                >
                  Первоначальный взнос
                </span>
              </div>
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
              >
                <button
                  className="Button Button_size_s Button_view_ghost Button_width_default Button_form_default Button_withIcon Button_onlyIcon MixFocus"
                  onClick={[Function]}
                  title="Опции"
                  type="button"
                >
                  <span
                    className="Icon Icon_size_xs IconKebab Button-Icon"
                  >
                    <svg
                      className="Icon-Svg"
                      viewBox="0 0 12 12"
                    >
                      <path
                        d="M7 2a1 1 0 11-2 0 1 1 0 012 0zm0 4a1 1 0 11-2 0 1 1 0 012 0zm-1 5a1 1 0 100-2 1 1 0 000 2z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
            <div
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
              >
                <div
                  className={
                    Object {
                      "has": [Function],
                      "is": [Function],
                      "mix": [Function],
                      "split": [Function],
                      "state": [Function],
                      "toString": [Function],
                    }
                  }
                >
                  <label
                    className={
                      Object {
                        "has": [Function],
                        "is": [Function],
                        "mix": [Function],
                        "split": [Function],
                        "state": [Function],
                        "toString": [Function],
                      }
                    }
                    data-testid="label"
                  >
                    Значение разового платежа
                  </label>
                  <div
                    className={
                      Object {
                        "has": [Function],
                        "is": [Function],
                        "mix": [Function],
                        "split": [Function],
                        "state": [Function],
                        "toString": [Function],
                      }
                    }
                  >
                    <div
                      className={
                        Object {
                          "has": [Function],
                          "is": [Function],
                          "mix": [Function],
                          "split": [Function],
                          "state": [Function],
                          "toString": [Function],
                        }
                      }
                    >
                      <div
                        className="TextField TextField_size_s TextField_view_default TextField_form_default TextField_width_full TextField_type_text"
                        data-testid="input"
                        onKeyDown={[Function]}
                      >
                        <input
                          autoFocus={false}
                          className="TextField-Input"
                          id="article_oneTimePaymentValue_1"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          placeholder="Значение"
                          type="text"
                          value=""
                        />
                        <div
                          className="TextField-Side TextField-Side_position_right TextField-Side_type_string"
                          title="rub."
                        >
                          rub.
                        </div>
                      </div>
                    </div>
                    <div
                      className={
                        Object {
                          "has": [Function],
                          "is": [Function],
                          "mix": [Function],
                          "split": [Function],
                          "state": [Function],
                          "toString": [Function],
                        }
                      }
                    >
                      <button
                        className="Button Button_size_s Button_view_ghost Button_width_default Button_form_default Button_withIcon Button_onlyIcon MixFocus"
                        onClick={[Function]}
                        title="Опции"
                        type="button"
                      >
                        <span
                          className="Icon Icon_size_xs IconKebab Button-Icon"
                        >
                          <svg
                            className="Icon-Svg"
                            viewBox="0 0 12 12"
                          >
                            <path
                              d="M7 2a1 1 0 11-2 0 1 1 0 012 0zm0 4a1 1 0 11-2 0 1 1 0 012 0zm-1 5a1 1 0 100-2 1 1 0 000 2z"
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <button
          className="Button Button_size_s Button_view_ghost Button_width_default Button_form_default Button_withIcon MixFocus"
          onClick={[Function]}
          type="button"
        >
          <span
            className="Icon Icon_size_xs IconAdd Button-Icon Button-Icon_position_left"
          >
            <svg
              className="Icon-Svg"
              viewBox="0 0 12 12"
            >
              <path
                d="M6.5 2.5h-1v3h-3v1h3v3h1v-3h3v-1h-3v-3z"
              />
            </svg>
          </span>
          <span
            className="Button-Label"
          >
            Добавить группу затрат
          </span>
        </button>
      </div>
    </form>
    <div />
  </div>
</div>
`;
